### ISSUE TITLE
Google::Protobuf.deep_copy not recursing for  sub-message on JRuby

### ISSUE THREAD

**Issue by NC-piercej** [protocolbuffers/protobuf#12505]
**What version of protobuf and what language are you using?**
Version: 3.22.1
Language: JRuby

**What operating system (Linux, Windows, ...) and version?**

Linux

**What runtime / compiler are you using (e.g., python version or gcc version)**

n/a

**What did you do?**

```
copy = Google::Protobuf.deep_copy(repeated_message_with_sub_messages)
copy[0].sub_message.object_id != repeated_message_with_sub_messages[0].sub_message.object_id
```

**What did you expect to see**
Sub message is a recursively deep-copied independent instance.

**What did you see instead?**
Sub message on the copy is the same as the original.

NOTE: issue still occurs even when an individual message is deep-copied

**Anything else we should know about your project / environment**
JRuby 9.4.2.0

**Comment by JasonLunn** [comment]
@NC-piercej - Is it feasible to provide a definition for `repeated_message_with_sub_messages` that fails? My first attempts to reproduce this locally on `jruby-9.3.4.0` weren't successful, so I want to make sure I'm testing this the way you are. Meanwhile I will install 9.4.2.0 to make sure this isn't JRuby version dependent.

**Comment by NC-piercej** [comment]
Here is a somewhat sanitized version still working locally:

```
enum Basis {
  PERCENT = 0;
  AMOUNT = 1;
}

message BasisValue {
  Basis basis = 1;
  double value = 2;
}

message Test {
  Basis basis = 1;
  double value = 2;
  SubMessage sub_message = 3;
}

message SubMessage {
  BasisValue example = 1;
}
```

Failing test script:

<img width="1033" alt="Screenshot 2023-04-19 at 4 29 22 PM" src="https://user-images.githubusercontent.com/17504011/233204345-bb49ef59-5df3-4c58-8414-144d9809dbd6.png">

**Comment by NC-piercej** [comment]
Can confirm that this appears to be fixed in `3.24`
